README.md
# Nagoriyuki

[![Gem Version](https://badge.fury.io/rb/nagoriyuki.svg)](http://badge.fury.io/rb/nagoriyuki)
[![Coverage Status](https://img.shields.io/coveralls/chibicco/nagoriyuki.svg)](https://coveralls.io/r/chibicco/nagoriyuki)
[![Build Status](https://travis-ci.org/chibicco/nagoriyuki.svg)](https://travis-ci.org/chibicco/nagoriyuki)

nagoriyuki(なごり雪) is a twitter/snowflake like ID generator.

This is the extraction of the minimum functionality with referred to the code of [drecom/barrage](https://github.com/drecom/barrage).

And there are only a few feature additions.

## Installation

Add this line to your application's Gemfile:

    gem 'nagoriyuki'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install nagoriyuki

## Usage

### Example

```ruby
# 39bit: msec (17.4 years from offset_epoch)
# 16bit: pid
# 9bit:  sequence
require 'nagoriyuki'

nagoriyuki = Nagoriyuki.new(
  "generators" => [
    {"name" => "msec", "length" => 39, "offset_epoch" => 1546268400000},
    {"name" => "pid", "length" => 16},
    {"name" => "sequence", "length" => 9}
  ]
)

id = nagoriyuki.next
# => 596299014690041345
  
ts = nagoriyuki.timestamp(id)
# => 1564039495

Time.at(ts)
# => 2019-07-25 16:24:55 +0900
```

### Generators

#### msec

#### pid

#### sequence

### Creating your own generator

```ruby
module Nagoriyuki::Generators
  class YourOwnGenerator < Base
    self.required_options += %w(your_option_value)
    def generate
      # generated code
    end

    def your_option_value
      options["your_option_value"]
    end
  end
end

nagoriyuki = Nagoriyuki.new("generators" => [{"name"=>"your_own", "length" => 8, "your_option_value"=>"xxx"}])
```

## License

[Copyright (c) 2014 Drecom Co., Ltd.](https://github.com/drecom/barrage/blob/master/LICENSE.txt)


CHANGELOG.md
# CHANGELOG

## 0.0.2

- Add Gemfile.lock to gitignore [#3](https://github.com/chibicco/nagoriyuki/pull/3)
- Bump json from 2.2.0 to 2.3.1 [#2](https://github.com/chibicco/nagoriyuki/pull/2)
- Bump activesupport from 5.2.3 to 6.0.3.1 [#1](https://github.com/chibicco/nagoriyuki/pull/1)

## 0.0.1 - First Release


spec/spec_helper.rb
$LOAD_PATH.unshift File.expand_path("../../lib", __FILE__)
require "bundler/setup"
require "rspec/its"
require "delorean"

require "coveralls"
Coveralls.wear!

require "nagoriyuki"

RSpec.configure do |config|
  include Delorean
  config.filter_run focus: true
  config.run_all_when_everything_filtered = true
end


spec/nagoriyuki_spec.rb
require "spec_helper"

describe Nagoriyuki do
  it "has a version number" do
    expect(Nagoriyuki::VERSION).not_to be nil
  end

  context "when initialized" do
    subject { Nagoriyuki.new(options) }

    context "with empty generators" do
      let(:options) { {"generators" => []} }
      its(:generators) { is_expected.to be_empty }
    end

    context "with generators" do
      let(:options) {
        {
          "generators" => [
            {"name" => "msec", "length" => 39, "offset_epoch" => 1396278000000},
            {"name" => "pid", "length" => 16},
            {"name" => "sequence", "length" => 9}
          ]
        }
      }
      it { expect(subject.generators.size).to eq(options["generators"].size) }
      its(:generators) {
        is_expected.to contain_exactly(
          be_kind_of(Nagoriyuki::Generators::Msec),
          be_kind_of(Nagoriyuki::Generators::Pid),
          be_kind_of(Nagoriyuki::Generators::Sequence)
        )
      }

      describe "#generate" do
        let(:nagoriyuki) { described_class.new(options) }
        subject { nagoriyuki.generate }
        it { is_expected.to be_kind_of(Integer) }
        it { is_expected.to be < 2 ** nagoriyuki.length }
      end

      describe "#length" do
        subject { described_class.new(options).length }
        it { is_expected.to eq(options["generators"].map { |h| h["length"]}.inject(:+)) }
      end

      describe "#timestamp" do
        let(:nagoriyuki) { described_class.new(options) }

        it {
          t = Time.parse("2019-01-01 00:00:00")
          id = time_travel_to(t) { nagoriyuki.generate }
          timestamp = time_travel_to(t) { nagoriyuki.timestamp(id) }
          expect(timestamp).to eq(t.to_i)
        }
      end
    end
  end
end


spec/nagoriyuki/generators/msec_spec.rb
require "spec_helper"
require "nagoriyuki/generators/msec"

describe Nagoriyuki::Generators::Msec do
  context "When initialized" do
    subject { described_class.new(options) }

    context "with empty hash" do
      let(:options) { {} }

      it { expect { subject }.to raise_error(ArgumentError) }
    end

    context "with required options" do
      let(:options) { {"length" => 16, "offset_epoch" => (Time.now.to_f * 1000).round } }

      it { is_expected.to be_instance_of(Nagoriyuki::Generators::Msec) }
      its(:length) { is_expected.to eq(options["length"]) }
      its(:offset_epoch) { is_expected.to eq(options["offset_epoch"]) }
    end
  end

  describe "#generate" do
    subject { described_class.new(options).generate }
    let(:length) { 39 }
    let(:offset_epoch) { (Time.parse("2019-01-01 00:00:00").to_f*1000).round }
    let(:options) { {"length" => length, "offset_epoch" => offset_epoch} }

    it { is_expected.to be_instance_of(Fixnum) }
    it { is_expected.to be < 2 ** length }

    context "generate two numbers between 10 millisecond " do
      subject { described_class.new(options) }

      it "numbers difference that generated in interval 10 millisecond should exactly 10" do
        t = Time.parse("2019-01-01 00:00:00")
        first = time_travel_to(t) { subject.generate }
        second = time_travel_to(Time.at(t.to_i, 10000)) { subject.generate }
        expect(second - first).to eq(10)
      end
    end
  end
end


lib/nagoriyuki/generators/base.rb
require "nagoriyuki/generators"
require "active_support/core_ext/class/attribute"

class Nagoriyuki
  module Generators
    class Base
      attr_reader :options
      class_attribute :required_options, :available_options
      self.required_options = %w(length)
      self.available_options = []

      def initialize(options = {})
        if (missing = missing_required_options(options)) && !missing.empty?
          raise ArgumentError, "Missing Required options: #{missing.join(', ')}"
        end
        @options = options
      end

      def name
        options["name"]
      end

      def length
        options["length"]
      end

      def generate
        raise NotImplemented, "Please Override"
      end

      def current
        raise NotImplemented, "Please Override"
      end

      def smallest(timestamp)
        raise NotImplemented, "Please Override"
      end

      def biggest(timestamp)
        raise NotImplemented, "Please Override"
      end

      private

      def bit_fill
        (2 ** length - 1)
      end

      def missing_required_options(given_options)
        required_options.reject { |k| given_options.has_key?(k) }
      end
    end
  end
end


lib/nagoriyuki.rb
require "nagoriyuki/version"
require "nagoriyuki/generators"
require "nagoriyuki/generators/base"
require "nagoriyuki/generators/msec"
require "nagoriyuki/generators/pid"
require "nagoriyuki/generators/sequence"
require "active_support/core_ext/string/inflections"

class Nagoriyuki
  class InvalidOption < StandardError; end
  attr_reader :generators

  #
  # generators:
  #   - name: msec
  #     length: 40 # (2 ** 40 - 1) / 1000 / 365 / 24 / 60 / 60.0
  #     start_at: 1546268400000 # 2019-01-01 00:00:00 +0900 (msec)
  #   - name: pid
  #     length: 16
  #   - name: sequence
  #     length: 8
  def initialize(options = {})
    @options = options
    @lock = Mutex.new
    @generators = @options["generators"].map do |h|
      generator_name = h["name"]
      "Nagoriyuki::Generators::#{generator_name.classify}".constantize.new(h)
    end
  end

  def generate
    @lock.synchronize do
      shift_size = length
      @generators.inject(0) do |result, generator|
        shift_size = shift_size - generator.length
        result += generator.generate << shift_size
      end
    end
  end
  alias_method :next, :generate

  def current
    @lock.synchronize do
      shift_size = length
      @generators.inject(0) do |result, generator|
        shift_size = shift_size - generator.length
        result += generator.current << shift_size
      end
    end
  end

  def timestamp(id)
    generator = @generators.detect {|g| g.name == "msec" }
    shift_size = length - generator.length
    ((id >> shift_size) + generator.offset_epoch) / 1000
  end

  def smallest(timestamp)
    @lock.synchronize do
      shift_size = length
      @generators.inject(0) do |result, generator|
        shift_size = shift_size - generator.length
        result += generator.smallest(timestamp) << shift_size
      end
    end
  end

  def biggest(timestamp)
    @lock.synchronize do
      shift_size = length
      @generators.inject(0) do |result, generator|
        shift_size = shift_size - generator.length
        result += generator.biggest(timestamp) << shift_size
      end
    end
  end

  def length
    @generators.inject(0) {|sum, g| sum += g.length }
  end
end
